logstash-core/src/main/java/org/logstash/plugins/pipeline/AddressState.java

```diff
@@ -0,0 +1,75 @@
+package org.logstash.plugins.pipeline;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Class for representing the state of an internal address.
+ */
+public class AddressState {
+    private final String address;
+    private final Set<PipelineOutput> outputs = ConcurrentHashMap.newKeySet();
+    private volatile PipelineInput input = null;
+
+    AddressState(String address) {
+        this.address = address;
+    }
+
+    /**
+     * Add the given output and ensure associated input's receivers are updated
+     * @param output
+     * @return
+     */
+    public boolean addOutput(PipelineOutput output) {
+        return outputs.add(output);
+    }
+
+    public boolean removeOutput(PipelineOutput output) {
+        return outputs.remove(output);
+    }
+
+    public PipelineInput getInput() {
+        return input;
+    }
+
+    /**
+     * Assigns an input to listen on this address. Will return false if another input is already listening.
+     * @param newInput
+     * @return true if successful, false if another input is listening
+     */
+    public boolean assignInputIfMissing(PipelineInput newInput) {
```

Comment:
on atomicity: can we guarantee that two threads won't interleave calls to `AddressState#assignInputIfMissing()` and `AddressState#unassignInput()`, is it worth marking these two methods as synchronized? call-frequency of both should be ultra-low, so sync overhead would be minimal.

Reply:
Good point! This should definitely be synchronized. We cannot make that guarantee.